### PR TITLE
Password recovery WIP

### PR DESCRIPTION
**Pull request by ilakumar**

### CODE REVIEW

File: router/auth/index.js [hunk 1]
```diff
@@ -30,161 +31,276 @@ module.exports = function(app){
...
+
+  if (!token){
+    return res.json({
+      err: 'No password reset token given'
+    });
+  } else if (!email || !password){
+    return res.json({
+      err: 'Must supply an email and password for password reset'
+    });
+  } else if (!newpassword){
```
Comment (stackptr) [inline router/auth/index.js]: Restating this point to track on Github: not really necessary to have this endpoint take both a `password` and `newpassword`--the client can be responsible for ensuring these two passwords match and then sending the singular password along. By exposing the endpoint to two different passwords and tasking it with checking them, the server is redoing work that the UI should already take care of. It's rather subjective, I guess, but something to think about, regarding the division of labor between server and client

File: router/auth/index.js [hunk 2]
```diff
@@ -30,161 +31,276 @@ module.exports = function(app){
...
+  ResetPasswordCtrl.finishReset({
+    token: token,
+    email: email
+  }, function(err, user){
+    if (err){
+      res.json({
+        err: err
+      });
+    } else {
+      user.hashPassword(password, function(err, hash){
```
Comment (stackptr) [inline router/auth/index.js]: `hashPassword` should probably be done in a Ctrl function and not in the router. This was probably already happening before you touched this file, so no worries

**Comment by stackptr** [comment]
Actually, do make sure that for a new user, the registrationCode is still being set properly.

File: models/User.js [hunk 3]
```diff
@@ -23,7 +23,8 @@ var userSchema = new mongoose.Schema({
     default: false
   },
   verificationToken: String,
-  registrationCode: String,
+  registrationToken: String,
```
Comment (ilakumar) [inline models/User.js]: I changed registrationToken back to registrationCode, I must have accidentally changed that!

File: router/auth/index.js [hunk 4]
```diff
@@ -30,161 +31,276 @@ module.exports = function(app){
   var router = new express.Router();
 
   router.get('/logout', function(req, res){
-      req.logout();
-      res.json({
-        msg: 'You have been logged out'
-      });
+    req.logout();
```
Comment (ilakumar) [inline router/auth/index.js]: Ok! That makes sense, I think I might have fixed the indentation on this page while I was making changes-- I will be more careful next time.

File: router/auth/index.js [hunk 5]
```diff
@@ -30,161 +31,276 @@ module.exports = function(app){
...
-    if (numNumber == 0) {
-      return res.json({
-        err: 'Password must contain at least one number'
+    } else if (!data.studentCode && !data.volunteerCode){
+      res.json({
+        err: 'Invalid registation code'
       });
-    }
+    } else {
+      var user = new User();
```
Comment (ilakumar) [inline router/auth/index.js]: I didn't make any changes to that-- again the issue of having fixed the indentation of the whole page.

**Pull request merged**